napari/config/core.py

```diff
@@ -0,0 +1,708 @@
+
+1. the config is cleared and ``refresh()`` is called
+2. ``refresh()`` updates the empty config with values from dicts in the
+   `config.defaults` list (currently empty)
+3. ``refresh()`` then updates the config by calling ``collect()`` ...
+   a. ``collect()`` gathers settings from various places:
+      - ``collect_yaml()`` looks for *any* `.yaml` files in the
+         ``config.paths`` list.  Notably, this includes
+         ``~/.config/napari/napari.yaml`` and anything else in there.
+      - ``collec_env()`` then looks for any environment variables that begin
+         with ``NAPARI_``
+   b. finally, ``collect()`` calls `merge()` on all of those discovered dicts,
+      to yield one merged result and returns it to the ``refresh()`` function
+      (by default, dicts discovered later in the process take priority).
+4. back in the ``refresh()`` function, that final merged result is used to
+   update the global config dict.
+5. At that point, the ``napari.yaml`` in the repo is read, if one doesn't
+   already exist, a file is created at ``~/.config/napari/napari.yaml`` (with
+   ``ensure_file()``) that is an exact copy of the repo ``napari.yaml`` but
+   with everything commented out, as a way to show users what they can change.
+6. finally, ``update_defaults()`` is called using the settings in the
+   `napari.yaml` file, which merges the configs again, but gives priority to
+   already existing values in the config (i.e. ones that were overwritten by
+   the user or the environment)
+
+Writing
+-------
+
+To write settings to disk, the :func:`sync` function is used.  This
+synchronizes the current ``config`` dict to a file at
+``join(config.PATH, 'napari.yaml')``.  If that file has changed since the last
+sync, it also reads new values from that file into the config.  If *both*
+the file and the config have changed, it prioritizes values in the config.
+"""
+import ast
+import logging
+import os
+import sys
+import threading
+from collections import defaultdict
+from functools import wraps
+from types import TracebackType
+from typing import (
+    Any,
+    Callable,
+    DefaultDict,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    Union,
+)
+
+import yaml
+
+from ..utils.misc import StringEnum
+from .deprecations import check_deprecations
+from .utils import canonical_name, ensure_file, merge, update
+
+logger = logging.getLogger('napari.config')
+
+
+#: a list of paths that get searched for .yaml files in :func:`collect_yaml`
+paths = [
+    os.getenv("NAPARI_ROOT_CONFIG", "/etc/napari"),
+    os.path.join(sys.prefix, "etc", "napari"),
+    os.path.join(os.path.expanduser("~"), ".config", "napari"),
+    os.path.join(os.path.expanduser("~"), ".napari"),
+]
+
+
+#: The primary directory for user config files
+#:
+#: Defaults to ~/.config/napari
+#: May be set with the environment variable "NAPARI_CONFIG"
+PATH = os.path.join(os.path.expanduser("~"), ".config", "napari")
```

Comment:
What with versioning? By default there should be different locations for stable one and development one. Maybe even python version may be important?